Web Inspector: BackendDispatcher::CallbackBase should have a virtual destructor

rniwa opened the pull request:
#### 74af740dd35fd0ca455c36e7672201aa44e820b3
<pre>
Web Inspector: BackendDispatcher::CallbackBase should have a virtual destructor
<a href="https://bugs.webkit.org/show_bug.cgi?id=274440">https://bugs.webkit.org/show_bug.cgi?id=274440</a>

Reviewed by Chris Dumez.

Added the virtual destructor to BackendDispatcher::CallbackBase since it has a bunch of subclasses.

* Source/JavaScriptCore/inspector/InspectorBackendDispatcher.h:

Canonical link: <a href="https://commits.webkit.org/279071@main">https://commits.webkit.org/279071@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03dcea1f353d0d607fe3f56536c639c6fe764694

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52332 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31664 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4753 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55606 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3055 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38092 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2754 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42555 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1949 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54428 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29291 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45164 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23632 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26520 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2439 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1214 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/45681 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48420 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2589 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57202 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/51840 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27458 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2622 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49951 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28691 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45284 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49193 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29603 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/64147 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7682 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28436 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12140 "Found 15 new JSC stress test failures: stress/spread-non-array.js.no-llint, wasm.yaml/wasm/function-tests/float-sub.js.wasm-eager, wasm.yaml/wasm/function-tests/grow-memory-3.js.wasm-eager, wasm.yaml/wasm/function-tests/i32-load8-s.js.wasm-eager, wasm.yaml/wasm/function-tests/memory-grow-invalid.js.wasm-eager, wasm.yaml/wasm/function-tests/struct.js.wasm-eager, wasm.yaml/wasm/function-tests/trap-load-shared.js.wasm-eager, wasm.yaml/wasm/function-tests/trap-load.js.wasm-eager, wasm.yaml/wasm/function-tests/trap-store-shared.js.wasm-eager, wasm.yaml/wasm/js-api/extension-MemoryMode.js.wasm-eager-jettison ... (failure)") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/inspector/InspectorBackendDispatcher.h:
```diff
@@ -56,6 +56,7 @@ class JS_EXPORT_PRIVATE BackendDispatcher : public RefCounted<BackendDispatcher>
     class JS_EXPORT_PRIVATE CallbackBase : public RefCounted<CallbackBase> {
     public:
         CallbackBase(Ref<BackendDispatcher>&&, long requestId);
+        virtual ~CallbackBase() { }
```
is this actually necessary?  there are no `virtual` methods and none of the subclasses have any additional member variables.

currently Web Inspector only ever stores subclasses as the subclass itself (i.e. there's nothing like `HashMap<Foo, RefPtr<BackendDispatcher::CallbackBase>>` anywhere), since ultimately these are just ways of converting data into structured JSON payloads

if we really do want/need to do this then id argue that `sendSuccess` should also be marked `virtual` with everything else being `final` since that's all that the subclass should ever override

rniwa replied:
If that's the case, we can suppress the static analyzer warning instead.

dcrousso replied:
if it helps, building Web Inspector (in JSC) will run `Source/JavaScriptCore/inspector/scripts/codegen/cpp_generator_templates.py` to generate `WebKitBuild/Release/DerivedSources/JavaScriptCore/inspector/InspectorBackendDispatchers.h` which contains code that uses `BackendDispatcher::CallbackBase` (among other things)

FWIW, there shouldn't be very many uses of `BackendDispatcher::CallbackBase` as it only exists to support protocol commands marked `async` (you can search for that in `Source/JavaScriptCore/inspector/protocol/`)

geoffreygaren replied:
> there are no virtual methods 

This isn't really relevant. These are the relevant variables:

> none of the subclasses have any additional member variables

> currently Web Inspector only ever stores subclasses as the subclass itself

If subclasses never add data members (with non-trivial destructors), or all uses of 'delete' always delete a pointer whose static type is the most derived class, or both, then there's no need for a virtual destructor.

However, I don't think it's sufficient to say that these facts are true right now. The questions we need to answer are:
* The next time someone edits e.g. cpp_generator_templates.py, what's going to stop them from adding a data member to a subclass?
* The next time someone edits Web Inspector, what's going to stop them from deleting a pointer whose static type is not  the most derived class?

If we're not sure of the answers, then it's probably best to make the destructor virtual. We do have some exceptions to this rule for performance-critical cases, but I don't think this case is performance-critical.

geoffreygaren replied:
You could imagine answering the first question by adding a static_assert for std::is_trivially_destructible -- but unfortunately the base class is not trivially destructible.

rniwa replied:
These functions take `Ref<BackendDispatcher::CallbackBase>` so it appears to me that we'd need virtual destructor after all.
```
    void waitForNavigationToCompleteOnPage(WebPageProxy&, Inspector::Protocol::Automation::PageLoadStrategy, Seconds, Ref<Inspector::BackendDispatcher::CallbackBase>&&);
    void waitForNavigationToCompleteOnFrame(WebFrameProxy&, Inspector::Protocol::Automation::PageLoadStrategy, Seconds, Ref<Inspector::BackendDispatcher::CallbackBase>&&);
```

rniwa replied:
In fact, `respondToPendingPageNavigationCallbacksWithTimeout(HashMap<WebPageProxyIdentifier, RefPtr<Inspector::BackendDispatcher::CallbackBase>>& map)` uses a HashMap of `RefPtr<Inspector::BackendDispatcher::CallbackBase>` as well.

geoffreygaren replied:
Technically "none of the subclasses have any additional member variables" has not yet been disproven, but I think this is a good demonstration that trying to think out all the uses of a class is not practical, and we should just do the safe thing instead.

rniwa replied:
If I'm not mistaken, calling delete with a wrong type triggers an undefined behavior, which is not a good practice:
https://cplusplus.github.io/CWG/issues/2474.html

rniwa committed:
Web Inspector: BackendDispatcher::CallbackBase should have a virtual destructor
https://bugs.webkit.org/show_bug.cgi?id=274440

Reviewed by Chris Dumez.

Added the virtual destructor to BackendDispatcher::CallbackBase since it has a bunch of subclasses.

* Source/JavaScriptCore/inspector/InspectorBackendDispatcher.h:

Canonical link: https://commits.webkit.org/279071@main
